enterprise/backend/src/metabase_enterprise/serialization/v2/load.clj

```diff
@@ -22,29 +29,41 @@
               (try
                 (load-one! ctx dep)
                 (catch Exception e
-                  (if (and (= (:error (ex-data e)) ::not-found)
-                           (serdes/load-find-local dep))
+                  (cond
                     ;; It was missing but we found it locally, so just return the context.
+                    (and (= (:error (ex-data e)) ::not-found)
+                         (serdes/load-find-local dep))
                     ctx
-                    ;; Different error, or couldn't find it locally, so rethrow.
+
+                    ;; It's a circular dep, strip of probable cause and retry. This will store an incomplete version
+                    ;; of an entity, but this is not a problem - a full version is waiting to be stored up the stack.
+                    (= (:error (ex-data e)) ::circular)
+                    (do
+                      (log/debug "Detected circular dependency" (serdes/log-path-str dep))
+                      (load-one! (update ctx :expanding disj dep) dep circuit-breaker))
+
+                    :else
                     (throw e)))))]
       (reduce loader ctx deps))))
 
 (defn- load-one!
-  "Loads a single entity, specified by its `:serdes/meta` abstract path, into the appdb, doing some bookkeeping to avoid
-  cycles.
+  "Loads a single entity, specified by its `:serdes/meta` abstract path, into the appdb, doing some bookkeeping to
+  avoid cycles.
 
-  If the incoming entity has any dependencies, they are recursively processed first (postorder) so that any foreign key
-  references in this entity can be resolved properly.
+  If the incoming entity has any dependencies, they are recursively processed first (postorder) so that any foreign
+  key references in this entity can be resolved properly.
 
   This is mostly bookkeeping for the overall deserialization process - the actual load of any given entity is done by
   [[metabase.models.serialization/load-one!]] and its various overridable parts, which see.
 
-  Circular dependencies are not allowed, and are detected and thrown as an error."
-  [{:keys [expanding ingestion seen] :as ctx} path]
+  Error is thrown on a circular dependency, should be handled and retried at the caller. `modfn` is an optional
```

Comment:
It seems smelly to me to define `modfn` only in terms of a specific function it gets used with, rather than according to how it works in an abstract setting. It's also a pretty opaque name.

This could point to just using a flag instead, until you find another use-case. If you were to just take a flag, this function could then encapsulate the detail of stripping (disj) `path` from `expanding` as well.

Reply:
Right now logic is incapsulated in `load-deps!`, I don't think it's a problem, but I agree that describing only in terms of circuit breaking is not ideal.